KEYCLOAK-6781 Docs for upgrading server to Wildfly 13 and upgrading J…

mposolda opened the pull request:
…DG to 7.2

mposolda committed:
KEYCLOAK-6781 Docs for upgrading server to Wildfly 13 and upgrading JDG to 7.2

matthewhelmke commented on server_installation/topics/cache/eviction.adoc:
```diff
@@ -5,61 +5,31 @@ There are multiple different caches configured for {project_name}.
-This cache is created implicitly and has twice the configured size.
-There are also separate caches for user sessions, offline tokens, and login failures.  These caches are unbounded in size as well.
+This cache is created implicitly and has twice the configured size. There are some more caches for save the authorization data, keys etc.
+There are also separate caches for user sessions, offline tokens, login failures and other things. These caches are unbounded in size by default.
```
The rewrite below is better. Even better yet would be if we could fill in details instead of saying "and so on" or "and other things."

This cache is created implicitly and has twice the configured size. There are also caches for the authorization data, keys, and so on. In addition, there are also separate caches for user sessions, offline tokens, login failures, and other things. These caches are unbounded in size by default.

mposolda replied:
I've changed like this in my local fork and added some details. Although adding more details is always double-side coin as it means more documentation changes will be needed in the future when implementation changes. See my comment below. So I've used this:

This cache is created implicitly and has twice the configured size. The same applies for the `authorization` cache, which holds for
the authorization data. The `keys` cache holds data about external keys and doesn't need to have dedicated revisions cache. Rather
it has `expiration` explicitly declared on it, so the keys are forced to be periodically downloaded from external clients or
identity providers.

In addition, there are also separate caches for user sessions, client sessions, offline tokens, login failures, action tokens and other things.
These caches are unbounded in size by default.

matthewhelmke replied:
Perhaps we can create a list of all caches and put that list in an obvious place in the documentation, then refer to it as the canonical reference when we make comments like "and other things".

My fear is that we don't want to frustrate readers/users by being vague. I definitely also agree with you about the double-sided coin of more detail means more documentation changes. The balance is that we should provide the information our users need, and only that information when we are writing procedures, but we should also find a way to provide a complete reference somewhere to the other details that they may eventually need.

mposolda replied:
Yes, I see. I agree with you that it is sometimes not easy to find the right balance.. In the end, I've added even more details into the "evictions" sections related to caches (talked today with @pedroigor about some things, which we're missing the docs). Also all the caches are used here and referenced from "replication" chapter.

matthewhelmke replied:
I like that approach, @mposolda. What do you think about replacing "and other things" with a reference like "and others, for more information about all of the caches, see ..."?

matthewhelmke commented on server_installation/topics/cache/replication.adoc:
```diff
@@ -1,7 +1,8 @@
 
-The `sessions`, `authenticationSessions`, `offlineSessions` and `loginFailures` caches are the only caches that may perform replication.  Entries are
+There are some caches like `sessions`, `authenticationSessions`, `offlineSessions`, `loginFailures` and few others, which are configured
+as distributed caches in case of clustered setup.  Entries are
```
Again, if details are available, it would be nice to replace "and a few others" with those details. If no, this rewrite will be sufficient.

There are caches like `sessions`, `authenticationSessions`, `offlineSessions`, `loginFailures` and a few others, which are configured as distributed caches when using a clustered setup. Entries are...

mposolda replied:
@matthewhelmke Well, I am not sure. Explicitly mention all the caches is more clear, but it has also one disadvantage.  Always when we add new cache into the configuration, we will also need to update this documentation with all the references to cache names. If you write "and few others", you explicitly mention to admin that there are more caches and he will quickly figure out by looking  to the configuration file what the caches are... :) WDYT? For now, I've just used the sentence you mentioned with "a few others" in it.

matthewhelmke replied:
See my comment above. It is not always an easy balance, and you probably know better than I what our users must know versus what they would like to have as reference.

matthewhelmke commented on server_installation/topics/cache/replication.adoc:
```diff
@@ -23,3 +24,7 @@ Here we've changed it so at least two nodes will replicate one specific user log
+TIP:  It is very good if you configure your environment to use loadbalancer with sticky sessions. It is beneficial for performance
+      as {project_name} server, where the particular request is served, will be usually the owner of the data from distributed cache and
+      hence it will be able to lookup the data locally. More details in <<sticky-sessions>>.
+
```
TIP:  It is generally wise to configure your environment to use loadbalancer with sticky sessions. It is beneficial for performance as {project_name} server, where the particular request is served, will be usually the owner of the data from the distributed cache and will therefore be able to look up the data locally. See <<sticky-sessions>> for more details.

matthewhelmke commented on server_installation/topics/clustering/troubleshooting.adoc:
```diff
@@ -1,19 +1,22 @@
-Note that when you run cluster, you should see message similar to this in the log of both cluster nodes:
-
+* Note that when you run cluster, you should see message similar to this in the log of both cluster nodes:
++
```
...run a cluster...

matthewhelmke commented on server_installation/topics/clustering/troubleshooting.adoc:
```diff
@@ -1,19 +1,22 @@
 For more information, see link:{appserver_jgroups_link}[JGroups] in the _{appserver_jgroups_name}_.
+
+* If you are interested about failover support (hight availability), evictions, expiration and cache tunings, you can go to the
+<<cache-configuration>>.
```
* If you are interested in failover support (high availability), evictions, expiration and cache tuning, see ...

matthewhelmke commented on server_installation/topics/operating-mode/crossdc.adoc:
```diff
@@ -233,7 +235,12 @@ Details of this more-detailed setup are out-of-scope of the {project_name} docum
-. Newer {jdgserver_name} releases require authorization before accessing protected caches over network. {project_name} requires updates to
+. Some {jdgserver_name} server releases require authorization before accessing protected caches over network.
++
+NOTE: You shouldn't see any issue if you use recommended {jdgserver_name} {jdgserver_version} server and this step can (and should) be ignored.
```
s/shouldn't/should not

matthewhelmke commented on upgrading/topics/keycloak/changes.adoc:
```diff
@@ -1,5 +1,25 @@
+  * If you use cross-dc integration, there is a need to add property `protocolVersion` with value `2.6` on `remote-store` . This is because
+  hotrod client ({project_name} server) is using infinispan 9 based on newer version of hotrod protocol than {jdgserver_name} server, which
+  is still using infinispan 8.
+
```
=== Migrating to 4.1.0

==== Upgrade to Wildfly 13

The {project_name} server was upgraded to use Wildfly 13 as the underlying container. This does not directly involve any specific {project_name} server functionality, but there are few changes related to the migration, which worth mentioning.

Dependency updates::
  The dependencies were updated to the versions used by Wildfly 13 server. For example, Infinispan is now 9.2.4.Final and Undertow is 2.0.9.Final

Configuration changes::
  There are few configuration changes in the `standalone(-ha).xml` and `domain.xml` files. You should follow the <<_install_new_version>> section
  to handle the migration of configuration files automatically. If more detail is needed, because, for example, you did some configuration
  changes on your own, here is a list of the most important changes:
  * Element `eviction` on infinispan caches is now deprecated (unused) and was replaced by element `object-memory`
  * There is no `jndi-attribute` on `cache-container` element in infinispan subsystem.
  * If you use cross-dc integration, you must add property `protocolVersion` with value `2.6` on `remote-store` . This is because
  the hotrod client ({project_name} server) is using infinispan 9 based on newer version of hotrod protocol than {jdgserver_name} server, which
  is still using infinispan 8.

matthewhelmke reviewed (changes requested):
Looks pretty good, just requesting some clean up.

mposolda committed:
KEYCLOAK-6781 Wildfly 13: more documentation changes

mposolda commented:
@matthewhelmke Thanks again for the review! I've updated the documentation per your feedback. But I also needed to do some updates to the content itself due the updates in the Wildfly 13 integration, which I needed to change in the meantime in the implementation PR: https://github.com/keycloak/keycloak/pull/5293 . I also added some more details to the "eviction" chapter as we discussed.

So new review will be probably needed. 

I've did all the changes into the separate second commit. Maybe it's easier for you to review it better to see all the new changes.

matthewhelmke commented:
@mposolda, thank you. I will do another review later today.

matthewhelmke commented on server_installation/topics/cache/eviction.adoc:
```diff
@@ -32,4 +35,24 @@ subsystem, which looks similar to this:
+If they are bounded, it would then be possible that some sessions will be lost. Expired sessions are cleared internally
+by {project_name} itself to avoid growing the size of these caches
+without limit. If you see memory issues due the big amount of sessions, you can try to:
+
```
If you see memory issues due to a large number of sessions...

matthewhelmke commented on server_installation/topics/cache/eviction.adoc:
```diff
@@ -32,4 +35,24 @@ subsystem, which looks similar to this:
+
+* Increase the memory for {project_name} server process
+
+* Lower the amount of owners to ensure that caches are saved just on single place. See <<_replication>> for more details
```
* Decrease the number of owners to ensure that caches are saved in one single place.

matthewhelmke commented on server_installation/topics/cache/eviction.adoc:
```diff
@@ -32,4 +35,24 @@ subsystem, which looks similar to this:
+usability for end users. See link:{adminguide_timeouts_link}[{adminguide_timeouts_name}] for more details.
+
+There is replicated cache `work`, which is mostly used to send messages among cluster nodes and it is also unbounded by default. However
+this cache should not cause any memory issues as entries in this cache are very short lived.
```
There is an additional replicated cache, `work`, which is mostly used to send messages among cluster nodes; it is also unbounded by default. However, this cache should not cause any memory issues as entries in this cache are very short-lived.

matthewhelmke reviewed (changes requested):
I have some very minor edits in your changes. Overall, I like these changes very much!

mposolda commented:
@matthewhelmke I've finally updated per your feedback. Sorry it took so long... :(

matthewhelmke reviewed (approved):
Thank you!
